File: C++/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size(), cols = rows > 0 ? grid[0].size() : 0;
        
        int islands = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    islands++;
                    paintGrid(grid, i, j, rows, cols);
                }
            }
        }
        return islands;
    }
    
    void paintGrid(vector<vector<char>>& grid, int i, int j, int rows, int cols) {
        if (grid[i][j] == '1') {
            grid[i][j] = '2';
            if (i > 0) paintGrid(grid, i - 1, j, rows, cols);
            if (j > 0) paintGrid(grid, i, j - 1, rows, cols);
            if (i < rows - 1) paintGrid(grid, i + 1, j, rows, cols);
            if (j < cols - 1) paintGrid(grid, i, j + 1, rows, cols);
        }
    }
};

File: Go/number-of-islands.go
func numIslands(grid [][]byte) int {
    count := 0
    for i := range grid {
        for j := range grid[i] {
            if grid[i][j] == '1' {
                count++
                paint(grid, i, j)
            }
        }
    }
    return count
}

func paint(grid [][]byte, i, j int) {
    if i < 0 || j < 0 || i >= len(grid) || j >= len(grid[0]) {
        return
    }
    if grid[i][j] != '1' {
        return
    }
    grid[i][j] = '0'
    paint(grid, i - 1, j)
    paint(grid, i + 1, j)
    paint(grid, i, j - 1)
    paint(grid, i, j + 1)
}